### ISSUE TITLE
Incremental build via Scons not working with MSVC

### ISSUE THREAD

**Issue by logicor-cn** [godotengine/godot#25403]
**Godot version:**

3.0.6-stable

**OS/device including version:**

Windows 10 1803
Python 3.6.4
pywin32 224
scons 3.0.4
Visual Studio 2015 14.0.25431.01 Update 3

**Issue description:**

Running the same scons build command does lengthy full build twice.

**Steps to reproduce:**

- Clone `3.0.6-stable` from git repository
- Setup build environment as above
- Run the following command line from `VS2015 x64 Native Tools Command Prompt`
```sh
scons -j8 platform=windows tools=yes target=release_debug
```

**Minimal reproduction project:**

**Comment by akien-mga** [comment]
Can you try with Scons 3.0.1? There have been various issues with 3.0.2+, this might be another regression.

**Comment by logicor-cn** [comment]
Switching to Scons 3.0.1 solved the problem.

Many thanks for the timely response!

**Comment by akien-mga** [comment]
Reopening as this needs to be fixed upstream, so it's good if we have a bug report about it for them.

**Comment by mwichmann** [comment]
`--debug=explain` might give some hint what scons is deciding (the `explain` function is, ummm, imperfect, so it also might not)

**Comment by Andrettin** [comment]
@akien-mga Thanks for letting me know :)

@mwichmann I tried what you said here, and it looks like Scons is mistakenly marking unchanged files as changed. One example:

```
[  6%] scons: rebuilding `thirdparty\bullet\BulletCollision\CollisionDispatch\btCollisionObject.windows.opt.tools.32.obj' because:
           `thirdparty\bullet\BulletCollision\CollisionDispatch\btCollisionObject.cpp' changed
           `thirdparty\bullet\BulletCollision\CollisionDispatch\btCollisionObject.h' changed
           `thirdparty\bullet\LinearMath\btSerializer.h' changed
           `thirdparty\bullet\LinearMath\btAlignedAllocator.h' changed
           `thirdparty\bullet\LinearMath\btAlignedObjectArray.h' changed
           `thirdparty\bullet\LinearMath\btMotionState.h' changed
           `thirdparty\bullet\LinearMath\btTransform.h' changed
           `thirdparty\bullet\LinearMath\btHashMap.h' changed
           `thirdparty\bullet\LinearMath\btScalar.h' changed
           `thirdparty\bullet\LinearMath\btMatrix3x3.h' changed
           `thirdparty\bullet\LinearMath\btQuaternion.h' changed
           `thirdparty\bullet\LinearMath\btVector3.h' changed
           `thirdparty\bullet\LinearMath\btQuadWord.h' changed
           `thirdparty\bullet\LinearMath\btMinMax.h' changed
           `C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\bin\HostX86\x86\cl.EXE' changed
```

It seems to be marking each and every source file as having changed.

**Comment by Andrettin** [comment]
@bdbaddog I haven't changed the Decider function at all (I didn't even know what it was before seeing your post and looking it up :P )

**Comment by bdbaddog** [comment]
Can you try removing your .sconsign file and see if it's still rebuilding when it shouldn't?
(Wondering if somehow we broke compat for .sconsigns between 3.0.1 and 3.0.3)

**Comment by Andrettin** [comment]
@bdbaddog I deleted the .sconsign.dblite file (I supposed that's what you meant since I found no .sconsign one), and now I get the following sort of output with --debug=explain:

```
[  2%] Retrieved `drivers\gles3\shaders\particles.glsl.gen.h' from cache
[  2%] scons: Cannot explain why `platform\windows\os_windows.windows.opt.tools.32.obj' is being rebuilt: No previous build information found
```

**Comment by bdbaddog** [comment]
can you run with --taskmastertrace=trace.log and pastebin it soemwhere.. though it's quite big.. maybe gzip and attach to this issue?

**Comment by mwichmann** [comment]
@Andrettin that's as expected (_No previous build information_), as all the signatures were removed. Now, if you try to build again after this run (starting from zero), does that still seem to be rebuilding things it shouldn't need to?

**Comment by Andrettin** [comment]
@mwichmann I just tried that; unfortunately, yes.

**Comment by bdbaddog** [comment]
@Andrettin can you do a clean build, then rebuild and capture the taskmaster trace file as described above?

**Comment by bdbaddog** [comment]
Is this only an issue with msvc? (linux builds?)

**Comment by akien-mga** [comment]
> Is this only an issue with msvc? (linux builds?)

Incremental builds seem to work fine on Linux with Scons 3.0.4 for me.

**Comment by Andrettin** [comment]
@bdbaddog Here is the taskmaster trace log file after doing a clean rebuild:
[trace.log](https://github.com/godotengine/godot/files/2809205/trace.log)

And here is the file after doing a normal build after that:
[trace.log](https://github.com/godotengine/godot/files/2809235/trace.log)

**Comment by bdbaddog** [comment]
@Andrettin - I'll take a look today/tomorrow.

**Comment by bdbaddog** [comment]
@Andrettin - Looks like those two files are identical?  Do you see a sconsign file?  Can you try running with python 2.7.x ?

**Comment by bdbaddog** [comment]
Are you using TryLink() from scons configure logic?

**Comment by bdbaddog** [comment]
O.k. The issue is triggered by some changes in the MD5-timestamp decider, or at least if I comment it out the build works fine (for small test).

Can someone else comment the Decider() line and see if you get proper incremental build with 3.0.4?

Meanwhile I'll see if I can figure out how none of our tests caught whatever issue there is with the new logic for MD5-timestamp. 

The change was to address SCons's issue 2980.
It seems like it's only windows. But not only py3.5+.


**One important note:**
I would never suggest doing a release build with these options.
So please have these conditionally set if not release.
In both cases you're reducing the certainty that it doesn't need to rebuild in exchange for a faster build.
Though this reduction in certainty is small, for release builds, I wouldn't advise it.
From SConstruct
```
# To decide whether to rebuild a file, use the MD5 sum only if the timestamp has changed.
# http://scons.org/doc/production/HTML/scons-user/ch06.html#idm139837621851792
env_base.Decider('MD5-timestamp')
# Use cached implicit dependencies by default. Can be overridden by specifying `--implicit-deps-changed` in the command line.
# http://scons.org/doc/production/HTML/scons-user/ch06s04.html
env_base.SetOption('implicit_cache', 1)
```

**Another note regarding requiring running in the VC developer command prompt**
Reviewing your build instructions you're telling builders that they need to run from within VC's command prompt.  SCons should detect everything that is needed for all supported versions of VC/VS.  If it's not, can you please bring the issue(s) to the scons users mailing list.  We'll resolve them and then you can remove that requirement...

**Comment by IronicallySerious** [comment]
After downgrading to Scons 3.0.1, running `scons platform=windows vsproj=true -Q --debug=explain` helped fix _something_.
There is some route that got triggered using this command, which sent this to the output:

``` 
Found MSVC version 14.1, arch amd64, bits=64
YASM is necessary for WebM SIMD optimizations.
WebM SIMD optimizations are disabled. Check if your CPU architecture, CPU bits or platform are supported!
[ 99%] scons: building `progress_finish' because it doesn't exist
[100%] progress_finish(["progress_finish"], [])
[100%]
```
After the `progress_finish` message popped up, Scons started reusing the obj files from previous builds

**Comment by akien-mga** [comment]
> **One important note:**
> I would never suggest doing a release build with these options.
> So please have these conditionally set if not release.
> In both cases you're reducing the certainty that it doesn't need to rebuild in exchange for a faster build.
> Though this reduction in certainty is small, for release builds, I wouldn't advise it.
> From SConstruct
> 
> ```
> # To decide whether to rebuild a file, use the MD5 sum only if the timestamp has changed.
> # http://scons.org/doc/production/HTML/scons-user/ch06.html#idm139837621851792
> env_base.Decider('MD5-timestamp')
> # Use cached implicit dependencies by default. Can be overridden by specifying `--implicit-deps-changed` in the command line.
> # http://scons.org/doc/production/HTML/scons-user/ch06s04.html
> env_base.SetOption('implicit_cache', 1)
> ```

I'm OK with either making those conditional (e.g. only for `target=debug` builds), or outright removing them. Do we really gain much from using those options?

**Comment by bdbaddog** [comment]
Yes those options will speed up your incremental build by 1-10 seconds per
iteration, so that's generally worth it when in "development" mode.
But for releases, better slower and less likely to run into any uncaught
build issues than shave a few seconds..

**Comment by IronicallySerious** [comment]
Downgrading to scons 3.0.1 did fix the problem a while ago but only recently I noticed that it has come back to not doing incremental builds.

My setup is:
>Windows,
VS 2017 (Community) (MSVC), 
Scons 3.0.1,
x64 bit version
Godot version: 3.2 master as of https://github.com/godotengine/godot/tree/86a74e2cb3f22b1d7f84f41cce673679d1b08a18

The scons starting logs could also help so I have put them below
```
1>------ Build started: Project: godot, Configuration: debug x64 ------
1>Starting SCons
1>**********************************************************************
1>** Visual Studio 2017 Developer Command Prompt v15.9.11
1>** Copyright (c) 2017 Microsoft Corporation
1>**********************************************************************
1>[vcvarsall.bat] Environment initialized for: 'x86_x64'
1>scons: Reading SConscript files ...
1>Configuring for Windows: target=debug, bits=default
1>Found MSVC compiler: x86_amd64
1>Compiled program architecture will be a 64 bit executable (forcing bits=64).
1>YASM is necessary for WebM SIMD optimizations.
1>WebM SIMD optimizations are disabled. Check if your CPU architecture, CPU bits or platform are supported!
1>Checking for C header file mntent.h... (cached) no
1>scons: done reading SConscript files.
1>scons: Building targets ...
```

**Comment by bdbaddog** [comment]
> Downgrading to scons 3.0.1 did fix the problem a while ago but only recently I noticed that it has come back to not doing incremental builds.
> 
> My setup is:
> 
> > Windows,
> > VS 2017 (Community) (MSVC),
> > Scons 3.0.1,
> > x64 bit version
> > Godot version: 3.2 master as of https://github.com/godotengine/godot/tree/86a74e2cb3f22b1d7f84f41cce673679d1b08a18
> 
> The scons starting logs could also help so I have put them below
> 
> ```
> 1>------ Build started: Project: godot, Configuration: debug x64 ------
> 1>Starting SCons
> 1>**********************************************************************
> 1>** Visual Studio 2017 Developer Command Prompt v15.9.11
> 1>** Copyright (c) 2017 Microsoft Corporation
> 1>**********************************************************************
> 1>[vcvarsall.bat] Environment initialized for: 'x86_x64'
> 1>scons: Reading SConscript files ...
> 1>Configuring for Windows: target=debug, bits=default
> 1>Found MSVC compiler: x86_amd64
> 1>Compiled program architecture will be a 64 bit executable (forcing bits=64).
> 1>YASM is necessary for WebM SIMD optimizations.
> 1>WebM SIMD optimizations are disabled. Check if your CPU architecture, CPU bits or platform are supported!
> 1>Checking for C header file mntent.h... (cached) no
> 1>scons: done reading SConscript files.
> 1>scons: Building targets ...
> ```
> 
> EDIT after 10 minutes: I am not very sure of what changes I made but incremental builds are occurring for me now. Strange.

If it starts happening again. Add --debug=explain --taskmastertrace=trace.log and post the output and the trace.log ile

**Comment by IronicallySerious** [comment]
@bdbaddog I got another full rebuild this time.
These are the logs after me making a few changes in 2-3 files (I did build Godot at least once before making these changes): [Output](https://pastebin.com/v5WEj5TZ)
These are the logs after I ran scons with the command that you provided: [Output](https://pastebin.com/r0ftYr14) [trace.log](https://drive.google.com/open?id=1heNyITetQJjWgj2NIQnMwytFYv_BA-VP)

**Comment by Anutrix** [comment]
Try 3.0.5

**Comment by Calinou** [comment]
I'll close this, as this is likely resolved in the latest SCons version (3.1.2). Latest SCons versions also supports Visual Studio 2019, which links Godot significantly faster.

**Comment by Xrayez** [comment]
There's another instance of incremental compilation issues as reported in #39117, and have preliminary pinned this down to the version of python used, no further idea why.
